certbot-nginx/certbot_nginx/parser.py

```diff
@@ -125,8 +148,20 @@ def get_vhosts(self):
                                         path)
                 vhosts.append(vhost)
 
+        self._update_vhosts_addrs_ssl(vhosts)
+
         return vhosts
 
+    def _update_vhosts_addrs_ssl(self, vhosts):
+        """Update a list of raw parsed vhosts to include global address sslishness
+        """
+        addr_to_ssl = self._build_addr_to_ssl()
+        for vhost in vhosts:
+            for addr in vhost.addrs:
+                addr.ssl = addr_to_ssl[addr.normalized_tuple()]
+                if addr.ssl:
+                    vhost.ssl = True
```

Comment:
```
server {
  server_name X;
  listen A;
  listen B;
  listen 443 ssl;
}

server {
  server_name Y;
  listen B;
  listen C;
}
```

This is the setup you're talking about?

A thing to note here is that `vhost.ssl` implies that at least one address in the server block is listening sslishly.

Whether or not address B (and therefore vhost Y) gets marked as `ssl` depends on what B is. If B is equivalent to `*:80`, it will get correctly marked as ssl, and pass that on to vhost Y. C will only be marked as `ssl` if it is also equivalent to B.